### PR TITLE
Backend queue integration

### PR DESCRIPTION
**Pull request by kalafut**
This was our WIP as of Friday

### CODE REVIEW

**Comment by briankassouf** [comment]
Were the vendor updates necessary, could we undo them?

**Comment by kalafut** [comment]
Sorry that all got added. I’ll clean it up.

File: builtin/logical/database/backend.go [hunk 1]
```diff
@@ -76,13 +76,60 @@ func Backend(conf *logical.BackendConfig) *databaseBackend {
...
 	b.logger = conf.Logger
 	b.connections = make(map[string]*dbPluginInstance)
+
+	// Always populate the queue to guard against nil pointers.
+	b.credRotationQueue = queue.NewTimeQueue()
+
+	// But only execute cred rotation if we're the leader.
+	// TODO what if leadership changes during the lifecycle of the application?
+	replicationState := conf.System.ReplicationState()
+	if replicationState.IsLeader() {
```
Comment (catsby) [inline builtin/logical/database/backend.go]: We also need to include a queue if this is a local mount, does this check include that?

File: builtin/logical/database/path_roles.go [hunk 2]
```diff
@@ -346,6 +353,25 @@ func (b *databaseBackend) pathRoleCreateUpdate() framework.OperationFunc {
...
+			// In case this is an update, remove any previous version of the item from the queue
+			if _, err := b.credRotationQueue.PopItemByKey(name); err != nil {
+				if _, ok := err.(*queue.ErrItemNotFound); !ok {
+					return nil, err
+				}
+			}
+
+			// Add their rotation to the queue
+			if err := b.credRotationQueue.PushItem(&queue.Item{
+				Key:      name,
```
Comment (catsby) [inline builtin/logical/database/path_roles.go]: I originally planned on just the `name` here as shown, but now I wonder how/if namespaces could affect this. Could namespaces be an issue?

File: builtin/logical/database/backend.go [hunk 3]
```diff
@@ -76,13 +76,60 @@ func Backend(conf *logical.BackendConfig) *databaseBackend {
...
+					}
+
+					if time.Now().Unix() > item.Priority {
+						// We've found our first item not in need of rotation
+						// TODO is this logic correct? Need to also check the logic creating the priority.
+						fmt.Printf("stmt: %+v\n", role.Statements.Rotation)
+						instance.SetCredentials(ctx, c, role.Statements.Rotation)
+						b.credRotationQueue.PushItem(&queue.Item{
+							Key:      item.Key,
+							Value:    role, // TODO is this what needs to be here?
```
Comment (catsby) [inline builtin/logical/database/backend.go]: This is as I intended, yes, however we could forgo a `Value` entirely and load the Role via the `item.Key` too, do you have thoughts?

File: builtin/logical/database/backend.go (hunk 1 above)
Comment (kalafut) [reply builtin/logical/database/backend.go]: I don't think local mounts are handled yet (@tyrannosaurus-becks can elaborate).

File: builtin/logical/database/backend.go (hunk 3 above)
Comment (kalafut) [reply builtin/logical/database/backend.go]: I can see that, and probably prefer the idea of the queue only helping to order events, with no data of its own.

File: builtin/logical/database/backend.go (hunk 1 above)
Comment (tyrannosaurus-becks) [reply builtin/logical/database/backend.go]: Yeah it checks them currently! However I haven't tested the logic IRL, it does need that. The logic checks if replication is disabled, which is how it's attempting to derive whether it's just a standalone. Not certain if that's what you mean by "local" mounts - if it's synonymous with "standalone" or if you also mean in a cluster but is the leader. The logic within the method definitely needs checking.

**Pull request merged**